### ISSUE TITLE
[Feature Request] Target for Rust Programming Language

### ISSUE THREAD

**Issue by ikarienator** [antlr/antlr4#1839]
I propose we have a target for Rust Programming Language since:

1. Rust has finally delivered stable release last year (currently version 1.16).
2. Rust has attracted more and more engineers recent years.

Rust has become a mature programming language with the constant effort from Mozilla. The I haven't found any good parser generators in the Rust world yet. Hopefully ANTLR can be one.

**Comment by mike-lischke** [comment]
This will only happen if someone from the commuity creates such a target and sends a pull request to the main ANTLR repo.

**Comment by neatnerd** [comment]
I would be willing to contribute, is somebody working on this subject already, so I dont do double work?

**Comment by KvanTTT** [comment]
You can also create a new topic in [ANTLR google group](https://groups.google.com/forum/#!forum/antlr-discussion)

**Comment by spease** [comment]
This would also be useful for a work project I'm working on.

**Comment by keppel2** [comment]
This would be great. A minimal implementation that omits some features would be fine (the Go implementation doesn't have a default visitor implementation for example). I just want to walk the tree myself.

**Comment by metadave** [comment]
I'd like to give this a shot unless someone else is already working on it.

**Comment by KvanTTT** [comment]
Yes, I think it's a great idea!

**Comment by xipheCom** [comment]
Would be awesome! Where can I watch the progress?

**Comment by rrevenantt** [comment]
@metadave I have done quite a lot of work on it recently. You can see it in https://github.com/rrevenantt/antlr4rust 
A decent bit of things are already working. If anybody wants to help i would gladly appreciate it.

BTW How complete target implementation should be to be merged in?

**Comment by mike-lischke** [comment]
@rrevenantt I recommend to use a clone of the antlr repo instead of an own repo, with an own structure. This way you can later create a pull request to get your code merged in and you can directly use the scripts, folders and test infrastructure of the main repo. Keep in mind that your target has to run the entire test suite that's executed for each of the targets.

**Comment by metadave** [comment]
@rrevenantt excellent! if that's the case, then perhaps I'll pick a different (new) target to work on, as I've wanted to learn all the internals of Antlr4 for quite some time.

**Comment by rrevenantt** [comment]
@mike-lischke Thanks. I was thinking about using git-subtree, so i have created it as a separate repo. Anyway I can easily merge them later. Nevertheless I am already using test infrastructure of the main repo, and quite some tests from test suit are already passing.

So the only requirement for PR is to pass complete runtime test-suit (and I suppose integrate api tests to CI) and then I can continue to improve it as part of main repo?

Is it better to ask subsequent questions on the mailing list as well?

**Comment by mike-lischke** [comment]
If that works out for then fine! I had some trouble when I worked on the C++ target, which also used an own repo, that's why I mentioned it. If a number of the tests already succeed then you are already pretty far. That's very impressive!

For further discussions I'd indeed recommend the mailing list.

**Comment by KvanTTT** [comment]
I like GitHub issues, I think it's also an appropriate place for feature/bug discussions.

**Comment by rrevenantt** [comment]
So I have implemented almost everything, less than 10 test are failing. Also there are some refactorings I would like to do beforehand. I think it will take couple more days.

**Comment by phodal** [comment]
It's really awesome.

**Comment by rrevenantt** [comment]
Finally, I have made it pass all tests. Those last parts were heavily based on inheritance and RTTI, which does not integrate with Rust paradigm after some point. So I had to rewrite parser generation for labeled alternatives to create enums. But it can generate almost fully idiomatic Rust syntax tree now. Was also going to remove most of the redundant allocations and cloning, but it appeared much harder than I expected. It is a real problem only in some rare corner cases, though, so I think Rust target can already be used.

Will still try to clean up and make a PR tomorrow. But I doubt that it will be merged quickly, also I will not be able to work on it much for about a week, so if anyone wants to use it ASAP, ping me and I will release it separately in my repo until it is merged.

**Comment by rrevenantt** [comment]
So while fixing that last issues, and preparing PR, I was thinking how everything is going to work if it gets merged in current state, and decided not to do it right now.
 - Firstly, the more I look at my current API design, the more I find ways to improve it ( if you know nothing about Rust, it is very different language in a lot of aspects, which forces you to design API differently). and they can break backward compatibility. Also feels like it needs much more tests than there currently is. And I have looked into other targets and it looks like they were merged in more finished state. 
 - Also it feels to me that this repository is not well suited for active development. And there is still quite a lot I can do. So I think that for everyone it would be more convenient, if I will continue to develop Rust target in my own repo for some more time. 

Nevertheless, I am not 100 % sure and if you prove me wrong, I can create PR next day. Meanwhile 
I have published first version on crates.io, so everyone is welcome to try it out and fill an issue on [rrevenantt/antlr4rust](https://github.com/rrevenantt/antlr4rust) if something is wrong.

**Comment by ericvergnaud** [comment]
Hi,

it's great that you've reached this state!

The code for this target might be merged anytime. However, the pace of releases is low: there was only 1 last year, and there was one just a month ago. 

We might afford a release for this, I'd say it's definitely worth it. But you probably want to increase confidence in your API before letting anyone build on the current one only to find that it breaks 3 months later (I haven't looked at the API so only suggesting that from your own sayings).

The recommended way of using ANTLR4 is through listeners or visitors, so make sure you get these right! Without inheritance, it's certainly a challenge!
 
Early adopters can always get started using your branch, but then they know what to expect in terms of API stability.

**Comment by mshijie** [comment]
Thank you for all your work. I've tried rust target in my project. It worked with my previous grammar file. What is the status of the rust target now?  When will it merge in？

**Comment by younes-io** [comment]
What is the status of the Rust traget please?

**Comment by TabulateJarl8** [comment]
antlr4rust hasn't had a commit in almost a year and still seems to only work for rust nightly, are there any forks or something that anyone is working on? I'd really like a rust target, and I would do it myself but I don't think I'm experienced enough in Rust to figure that out. Any updates would be great

**Comment by KvanTTT** [comment]
> Sorry for my absence, was really busy irl.

Maybe it's even better because I significantly simplified test infrastructure. Just compare test classes for C++ in the [old infrastructure](https://github.com/antlr/antlr4/tree/master/runtime-testsuite/test/org/antlr/v4/test/runtime/cpp) and in the [new infrastructure](https://github.com/antlr/antlr4/tree/dev/runtime-testsuite/test/org/antlr/v4/test/runtime/cpp). Now it's much simpler to integrate a new target.

**Comment by mike-lischke** [comment]
@KvanTTT Do we have a document describing how to integrate a new target into the test infrastructure? I vaguely remember having read instructions in the past (after all I integrated the C++ target), but that info is probably outdated now and I don't remember where I read it.

**Comment by KvanTTT** [comment]
Also, I'd recommend looking at other targets and implementing in the similar way. For instance, here is the setup for C++: https://github.com/antlr/antlr4/blob/dev/runtime-testsuite/test/org/antlr/v4/test/runtime/cpp/CppRunner.java

**Comment by ericvergnaud** [comment]
A significant potential benefit of a Rust target is... WebAssembly. My own experiments trying to move some bits of the js runtime to webasm have not been encouraging (the cost of crossing the js/webasm fence is higher than the benefits).  
Chances are that with a significant part of the runtime natively optimized for webasm, this balance will revert and we could use a rust originating webasm runtime to power the core of the js/ts one.

**Comment by parrt** [comment]
Guys...sorry but not sure I'd be interested in yet another platform to support / manage.

**Comment by khoshrou** [comment]
Seems Rust is growing well. Really waiting for Rus target.

**Comment by rkrishnasanka** [comment]
@rrevenantt I saw that the 0.3-prerelease got merged. I'm curious to know where I should be using your repo or this if I want to have a rust target.

**Comment by t-b** [comment]
@rrevenantt Friendly ping. Any update on your end? We are currently starting a ANTLR project and thinking about using your rust target. But it looks a bit stalled.

Just to be clear: It's totally fine if you don't continue this, just us knowing that would be helpful ;)

**Comment by sax** [comment]
Me, too! @rrevenantt I just started looking at your fork, but don't have a good sense of how much work it would be to refactor based on the latest dev branch. I might play around on my own, but if you had any thoughts I'd love to hear them.

**Comment by luzidchris** [comment]
> @rrevenantt Friendly ping. Any update on your end? We are currently starting a ANTLR project and thinking about using your rust target. But it looks a bit stalled.

I have the same impression - unfortunately. 

> Me, too! @rrevenantt I just started looking at your fork, but don't have a good sense of how much work it would be to refactor based on the latest dev branch. I might play around on my own, but if you had any thoughts I'd love to hear them.

It would be great if anybody had the time and skill to continue this. There are quite a few pull requests pending for that project. It's probably worth checking them first. It seems people have spend effort on creating patches to fix issues or realize additional functionality. It would probably make sense to go from there?

**Comment by fawdlstty** [comment]
Hello, there is new information for this functions? I have been looking forward to rust target for a long time

**Comment by alexsnaps** [comment]
I was myself in need for this, so decided to revive some of the work that @rrevenantt had started, the code is [here](https://github.com/antlr4rust/antlr4/tree/rust-target/runtime/Rust) and a [`v0.3.0-beta2` is published under a new crate on crates.io](https://crates.io/crates/antlr4rust). 

This was the minimal work to get my stuff to compile, I am now moving torwards making it all work for my usecase. More work is needed tho: 
 - Update antl4
 - Get rid of all warnings et al
 - more...?

**Comment by wakuflair** [comment]
Hi @alexsnaps , thank you for continuing work in the repo, a Rust target for ANTLR has been on my wish list for a long time.

About the name `antlr4rust` , it confuses me a bit -- it means "ANTLR for Rust" or "ANTLR v4 + Rust"? What happens if ANTLR goes to v5?

**Comment by alexsnaps** [comment]
@wakuflair The idea is to get this to a state where this can be upstreamed here. 
I "just" wanted to create an org where people interested in participating could. As wrt the crate, well... sure. Guess `antlr-rust` "might" be reusable now, if @rrevenantt is willing to "share" or "donate it". Now is anything roadmap'ed towards a v5 release? Would that warrant a "antlr5rust" crate anyways? I mean today, afaict, all antlr4 `artifactId`s are, well, all `4` as well, aren't they?

**Comment by cyqw** [comment]
@alexsnaps 
i have time to update antlr4 to latest

> I was myself in need for this, so decided to revive some of the work that [@rrevenantt](https://github.com/rrevenantt) had started, the code is [here](https://github.com/antlr4rust/antlr4/tree/rust-target/runtime/Rust) and a [`v0.3.0-beta2` is published under a new crate on crates.io](https://crates.io/crates/antlr4rust).
> 
> This was the minimal work to get my stuff to compile, I am now moving torwards making it all work for my usecase. More work is needed tho:
> 
> * Update antlr4
> * Get rid of all warnings et al
> * more...?

**Comment by kaby76** [comment]
I'd like to test the Rust target against the 369 grammars in grammars-v4. But it's inconvenient to build the tool. As Mike did with antlr4ng, can an "antlr4-rust-complete.jar", or whatever it should be called, containing the Antlr4 tool with Rust templates, be added to the Rust package itself https://crates.io/crates/antlr4rust? This will allow me to skip the git clone/mvn build of the tool and go directly into testing. This was the solution @mike-lischke did for [antlr4ng](https://www.npmjs.com/package/antlr4ng?activeTab=versions), and it worked. Note, since version 3.0.15, Mike removed the tool .jar from the runtime package because he rewrote the Antlr4 tool in full TypeScript in [antlr-ng](https://www.antlr-ng.org/). Also, you might consider adding the target to antlr-ng. It works great.

**Comment by rrevenantt** [comment]
>  The idea is to get this to a state where this can be upstreamed here. I "just" wanted to create an org where people interested in participating could. As wrt the crate, well... sure. Guess `antlr-rust` "might" be reusable now, if [@rrevenantt](https://github.com/rrevenantt) is willing to "share" or "donate it". Now is anything roadmap'ed towards a v5 release? Would that warrant a "antlr5rust" crate anyways? I mean today, afaict, all antlr4 `artifactId`s are, well, all `4` as well, aren't they?

As soon as it is upstreamed to the main `antlr` organisation, i am ready to share an ownership over `antlr-rust` crate with main antlr maintainers. And I would like to continue participating in this project after it is upstreamed, albeit in a quite limited fashion. 

However what we need to decide is whether we are going to make upstreaming pull request from @alexsnaps  repository or mine. It seems that you have changed project structure from being in a separate repository into a main antlr4 file tree. I am not sure that it is the best solution in case of Rust because separate repository has advantage of having separate CI and separate issue tracker for Rust specific issues. I think this is important because Rust target is the most different from other language targets so there can be a lot of issues and tests for these language related differences. Another advantage is ability to publish right from CI when new version number is pushed. And I have done it in github's CI for my repository. 

But I think here we need someone from main antlr maintainers to comment on what they prefer.

PS. IIRC initially it was impossible to publish a crate which is not located in root of git repository. That was my main motivation originally. But now that problem seems to be solved.

**Comment by alexsnaps** [comment]
>IIRC initially it was impossible to publish a crate which is not located in root of git repository

This may have been the case, but is not anymore indeed. We're doing this for some of our crates anyways. 
I don't see any issues with having it all within antlr over an external repo, but I could be wrong. Let's list pros & cons. 

wrt where to do that from, I think that's a commitment issue. If @rrevenantt you're up to reowning all of this, please do so. I mean you did most of the heavy lifting. I can certainly add you as a maintainer on the antlr4rust org if you're willing to share load there.

**Comment by alexsnaps** [comment]
> But it's inconvenient to build the tool.

It should just be a matter of running maven on the repo, but I can upload the jar somewhere. I'll do that today

**Comment by rrevenantt** [comment]
In my original repo there was a CI that built and published antlr tool on releases page. As for including tool in the crate itself i don't think it is possible because crate source has to correspond to a git repo of that crate.

**Comment by alexsnaps** [comment]
>As for including tool in the crate itself i don't think it is possible because crate source has to correspond to a git repo of that crate.

We could do it, shouldn't be an issue to have the jar (as I think that's what you mean) in there... but I wouldn't do it. Anyways... different questions.

**Comment by kaby76** [comment]
Maybe it is best to place the tool and runtime into separate crates and let Cargo manage the dependencies. While I could add a git clone/mvn build step in Github Actions for the Rust/Antlr4 tool, like [I did for antlr-ng](https://github.com/antlr/grammars-v4/blob/master/.github/workflows/main.yml#L153-L162), I would prefer a solution that doesn't tightly bind the version of the [.github/workflows/main.yml](https://github.com/antlr/grammars-v4/blob/master/.github/workflows/main.yml) file to a specific version of the runtime in the driver templates (https://github.com/antlr/grammars-v4/tree/master/_scripts/templates). If one places the clone/mvn build in the workflow, Dependabot cannot update the tool version because it does not parse and update git clone versions in the .yaml file. But, Dependabot does manage [Cargo dependencies](https://docs.github.com/en/code-security/dependabot/ecosystems-supported-by-dependabot/supported-ecosystems-and-repositories#cargo), and will offer an update to the tool and runtime versions when a new release is made, and avoids tool and runtime version skews.

**Comment by rrevenantt** [comment]
> wrt where to do that from, I think that's a commitment issue. If @rrevenantt you're up to reowning all of this, please do so. I mean you did most of the heavy lifting. I can certainly add you as a maintainer on the antlr4rust org if you're willing to share load there.

Not quite. Either way it will eventually end up under antlr organisation. But I think that given that you moved to a single repo, it mostly depends on desired project structure to be upstreamed. If it is preferred to have it in main source tree, then there is not much reason to keep my repos, and i can join yours. In opposite case i think it would be easier to add you to mine. 

And I see only one small disadvantage for having a separate repo: harder to synchronize changes that involve both repos. But others might have a different opinion.

> Maybe it is best to place the tool and runtime into separate crates and let cargo manage the dependency? 

I don't think it would be that simple. The problem is that tool is a build dependency and runtime is a regular dependency and cargo considers those as independent dependency trees. Although it might be possible to do something tricky like reading `Cargo.toml` in `build.rs` script

**Comment by kaby76** [comment]
>I don't think it would be that simple. The problem is that tool is a build dependency and runtime is a regular dependency and cargo considers those as independent dependency trees. Although it might be possible to do something tricky like reading Cargo.toml in build.rs script

That's right. This is *exactly* what I have to do for testing for most of the other targets! And this is why it's been a pain to get this all working. I wouldn't expect it to be any different for the Rust target because of the Antlr tool/runtime separation.

* Build a .jar for your tool and put it in a crate with a specific version.
* Write templates for Rust parser for trgen. The .toml file for testing would need to reference specific crates for the runtime and tool. 
* For testing:
   * Run trgen to instantiate a Rust driver parse app from templates.
   * Build:
       * Load crates.
        * <del>Use sed/awk/Powershell/whatever to grab the version of the runtime tool from the .toml file.</del> Just use "find . -name '*.jar'` to find the loaded Antlr Tool .jar.
        * Run `java -jar` with the appropriate .jar.
        * Proceed with a Rust compilation.

In the meantime, I'll work out the templates for a Rust parser driver.

**Comment by alexsnaps** [comment]
Ok, so there is some work that's been done in the [antlr4rust/antlr4 fork](https://github.com/antlr4rust/antlr4), and more to be done (I need to file issues 🤦 ). Tho what is there is pretty functional (the `cel-rust` parser now [uses the proper grammar](https://github.com/clarkmcc/cel-rust/pull/135)). 

I think getting antlr4 up to date is probably the most important step. Until then, and if proven wrong, I'd keep the rust runtime in that fork (which includes the "other" published [crate](https://crates.io/crates/antlr4rust))

@rrevenantt I'll add you to the org, which would let you be involved as you see fit. Any PR is more than welcome, and let's track issues there then, wdyt? Having the work laid out in this discussion seem suboptimal :P

**Comment by alexsnaps** [comment]
@kaby76 :
> As Mike did with antlr4ng, can an "antlr4-rust-complete.jar", or whatever it should be called, containing the Antlr4 tool with Rust templates, be added to the Rust package itself https://crates.io/crates/antlr4rust?

https://github.com/antlr4rust/antlr4/releases/tag/v0.3.0-beta2

**Comment by fawdlstty** [comment]
@alexsnaps Thank you for your efforts. This is an amazing job. It would be even better if the version number could follow the main line of ANTLR.

**Comment by cyqw** [comment]
I had merged master branch into rust-target branch on my fork. But the testcase can't build with error bellow. I can't known how many issues on my work. Could you help me?
@alexsnaps 
https://github.com/cyqw/antlr4/tree/rust-target
 

`error[E0782]: expected a type, found a trait
  --> D:\cy\git\antlr4\runtime\Rust\build.rs:46:21
   |
46 | ) -> Result<(), Box<Error>> {
   |                     ^^^^^
   |
help: you can add the `dyn` keyword if you want a trait object
   |
46 | ) -> Result<(), Box<dyn Error>> {`


> Ok, so there is some work that's been done in the [antlr4rust/antlr4 fork](https://github.com/antlr4rust/antlr4), and more to be done (I need to file issues 🤦 ). Tho what is there is pretty functional (the `cel-rust` parser now [uses the proper grammar](https://github.com/clarkmcc/cel-rust/pull/135)).
> 
> I think getting antlr4 up to date is probably the most important step. Until then, and if proven wrong, I'd keep the rust runtime in that fork (which includes the "other" published [crate](https://crates.io/crates/antlr4rust))
> 
> [@rrevenantt](https://github.com/rrevenantt) I'll add you to the org, which would let you be involved as you see fit. Any PR is more than welcome, and let's track issues there then, wdyt? Having the work laid out in this discussion seem suboptimal :P

**Comment by alexsnaps** [comment]
I won't provide any feedback here anymore, please let's move discussions on to the antlr4rust/antlr4 repo.
As with regards to "help" with getting it fork to be up to date with the upstream, which I think is what you mean @fawdlstty,  (for which I created an issue [here](https://github.com/antlr4rust/antlr4/issues/2), I'm sorry @cyqw , but that's the work to be done. Which I will try to carve out time to do so, but if in the meantime if you are trying to do this already, which is an effort I appreciate, I can't handhold you thru the process tho.

**Comment by khoshrou** [comment]
Is there any good news for Rust target? whats is the state of the Rust target?

**Comment by alexsnaps** [comment]
We just released the latest [tool](https://github.com/antlr4rust/antlr4/releases/tag/v0.5.0) and matching crate [`v0.5.0`](https://crates.io/crates/antlr4rust/0.5.0)... slowly making progress towards [opening a PR with the upstream 🤞 ](https://github.com/antlr4rust/antlr4/milestone/1).
A huge thank you to @cyqw for getting us to antlr v4.13.3-SNAPSHOT

**Comment by mike-lischke** [comment]
Hey @alexsnaps 👋  

Would you be interested in creating a target generator for the Rust target usable with [antlr-ng](https://github.com/antlr-ng/antlr-ng/tree/features/target-generators/src/default-target-generators)?

I’d be really happy to help you get started with the development setup — you’d just need TypeScript instead of String Template 4.  

You’d have full ownership of this target generator, and it doesn’t need to be integrated with antlr-ng right away. If you’d like, I can share more details and ideas about it!

**Comment by alexsnaps** [comment]
> Would you be interested in creating a target generator for the Rust target usable with [antlr-ng](https://github.com/antlr-ng/antlr-ng/tree/features/target-generators/src/default-target-generators)?

I'll be transparent here, and the answer is `no`. I was "only" looking for a working solution to use the [official CEL grammar](https://github.com/google/cel-go/blob/master/parser/gen/CEL.g4)... which led to all this work, which is more than I _really_ wanted to take on. So, selfishly, there is no reason for me to add yet more work on top of the never ending pile that I'm currently looking at 🤦 . So while I'm happy pushing this effort forward, I can't see how I could do a proper job at this stage at handling yet another stack. Sorry.

**Comment by mike-lischke** [comment]
Fair enough @alexsnaps. I had not much hope anyway and though I just ask. I could take your ST4 template and convert that myself, but I also don't want to manage any other target but TS.

Anyway, thanks for your contribution! I think that Rust target will be very welcome in the community.

**Comment by kaby76** [comment]
@alexsnaps Thanks for the info--expecially the code at https://github.com/google/cel-go/tree/master. That was the missing link I needed. The last time I tried, at the end of June, I couldn't get past a build. Documentation is sufficient for experts, but nothing beats a simple application with build and run scripts that lays it out, requiring no knowledge beyond basic Bash and the installation of Rust (https://github.com/kaby76/abnf_rust). And, while there's a lot I see missing from the Antlr Rust runtime API, I can now write templates to generate a Rust app for any Antlr grammar in grammars-v4 for testing. https://github.com/antlr/grammars-v4/issues/4514

**Comment by alexsnaps** [comment]
>Thanks for the info--expecially the code at https://github.com/google/cel-go/tree/master. That was the missing link I needed. The last time I tried, at the end of June, I couldn't get past a build. Documentation is sufficient for experts, but nothing beats a simple application with build and run scripts that lays it out, requiring no knowledge beyond basic Bash and the installation of Rust (https://github.com/kaby76/abnf_rust). 

Dunno if you "mislinked", but if you refer to antlr, CEL grammar and Rust, then that'd be [here](https://github.com/cel-rust/cel-rust/blob/master/cel/src/parser/gen/mod.rs)

>And, while there's a lot I see missing from the Antlr Rust runtime API, I can now write templates to generate a Rust app for any Antlr grammar in grammars-v4 for testing. [antlr/grammars-v4#4514](https://github.com/antlr/grammars-v4/issues/4514)

That's great! Whatever you end up identifying as missing/broken/whatever, please open issues in [our repo](https://github.com/antlr4rust/antlr4), trying to track remain work that needs to happen _before_ we open the PR here, in this [milestone](https://github.com/antlr4rust/antlr4/milestone/1)

Again, I'm not antlr4 expert by any means, I was "just" trying to get cel-rust up to speed.

**Comment by kaby76** [comment]
The Rust target looks pretty solid. There are some minor problems that should be easy to fix.

With that, I ran the Rust target v0.5.1 against the [c](https://github.com/antlr/grammars-v4/tree/master/c) and [java/java20](https://github.com/antlr/grammars-v4/tree/master/java/java20) grammars, and compared it to some other targets (/generator) (Antlr4ng, CSharp, Cpp, Dart, Go, Java, JavaScript, TypeScript, Python3). The java20 grammar has significant ambiguity and large k's. Both these grammars were chosen because they are "action-free". I didn't feel like writing any Rust ports.

The Rust target performs on par with the Cpp or Dart targets. See graphs below.

Scripts and data: [scripts-and-data.zip](https://github.com/user-attachments/files/23199661/scripts-and-data.zip). Sample size 5 runs of grouped parse (1st input file caching DFAs, rest of the inputs using the cached DFAs).

BTW, I wrote an Earley engine that uses the Antlr Tool generated epsilon-NFA/ATNs and then compared that with epsilon-free NFA/ATNs. The Earley parser with eps-free NFA/ATNs generally performs on par with the non-cached DFA (aka not "warmed up") ALL(star), sometimes faster.

### c
![Image](https://github.com/user-attachments/assets/6908028c-817a-4e12-9689-09bb8d75813f)

### java/java20

![Image](https://github.com/user-attachments/assets/50a42e23-d413-4281-97d6-3daa662ff49a)

**Comment by parrt** [comment]
Interesting!!!  Thanks Ken

Dictation in use. Please excuse homophones, malapropisms, and nonsense.

**Comment by alexsnaps** [comment]
Hey y'all, we (@rrevenantt - he's back 🙌 - and me) are trying to push during this upcoming holiday season to get a PR up here for the Rust runtime. There are [a few questions](https://github.com/antlr4rust/antlr4/issues/24#issuecomment-3669691119) we have identified, but there might be... unknown unknowns as one once said.

At best this is a heads up, but if someone knows of things we're missing or need doing prior to opening that PR, we'd appreciate if you'd let us know. Thanks! And thanks to @kaby76 for the help provided over the last few months that helped me (hopfully!) do the right thing.

**Comment by alexsnaps** [comment]
Based on the discussion on the fork, it seems like your preferred approach @parrt would be to keep the Rust support in a separate jar, is that a fair understanding? Trying to clear out misunderstandings here and align expectations. 

I totally get it that this would offset the maintenance burden away from you all here, which I totally understand. But I could be misunderstanding goals here... I guess the question is: how do you think we best go about adding Rust support on the long run?

**Comment by jimidle** [comment]
There are some ongoing discussions about Rust support, whereby I might be
sponsored to create a rust target, but it’s still up in the air. If
anti-aging changes, I will let you know. In the meantime, I wish you good
luck.

Jim

**Comment by alexsnaps** [comment]
@jimidle Thanks for the update. Until then, don't hesitate to reach out if needed.
I started cleaning the dependency to the `dev` branch here... Turning out to be fairly clean. I'll keep the fork up to date until a clear path forward emerges... the two jar approach might actually fairly easy to implement; so merging here too then - really is a question of this is best maintained. Which I totally understand 🙏

**Comment by tinovyatkin** [comment]
I've created my own Rust target runtime for ANTLR which beats built-in Go target by performance - https://github.com/ophi-dev/antlr-rust-runtime

**Comment by kaby76** [comment]
> I've created my own Rust target runtime for ANTLR which beats built-in Go target by performance - https://github.com/ophi-dev/antlr-rust-runtime

Performance depends on the grammar and input. For example, for [antlr4rust](https://github.com/antlr4rust/antlr4) and Go using the [java/java/ grammar](https://github.com/kaby76/grammars-v4/tree/g4-4924/java/java), and parsing the huge JDK SDK source (14207 .java files), the [Rust](https://github.com/antlr4rust/antlr4) port beat the Go port as well (https://github.com/kaby76/grammars-v4/blob/g4-4924/java/java/README.md#performance-summary-n3-runs-mean--sem). But, both Rust and Go were not as good as the CSharp or Java ports. (An explanation of the standardized performance measurements is at `https://github.com/kaby76/Trash/tree/main/src/trgen#summary-output.) In comparison, using a [much smaller test suite](https://github.com/antlr/grammars-v4/tree/master/java/java/examples), Dart is fastest, followed by Cpp, Rust, Go, Java, and CSharp.

I like what you are doing, but .interp files don't represent actions, and the `--grammar` doesn't seem to work. If you are using Claude, you may as well just port the entire Antlr tool to Rust (or other language) and output .interp files with actions, making an "interp" file that is complete to generate for any target language. Most top TIOBE grammars in grammars-v4 have actions and predicates because the parse depends on semantics, e.g., a symbol table used to make the right choice in the EBNF. The grammars-v4 repo does do testing against the [antlr4rust](https://github.com/antlr4rust/antlr4) target, but not for most grammars.

The main issues I have with  [antlr4rust](https://github.com/antlr4rust/antlr4) is that there are bugs in Rust.stg (e.g., the name of the parser with extra "Parser"), and the API, e.g., where the context (parse tree) is not available For example, for the [c grammar](https://github.com/antlr/grammars-v4/tree/master/c), semantic predicates have to look backwards in the token stream as opposed to looking at the current parse tree up to the current token, an inconvinient trade off.

**Comment by tinovyatkin** [comment]
@kaby76 you're right on the first point and I'll own it — "beats Go" was too glib. Performance is grammar- and input-dependent, and your JDK-scale numbers showing antlr4rust and Go both trailing C#/Java on `java/java` are a good reminder that the ranking flips with the workload. What I can defend is narrower and reproducible: on the fixtures in [`tools/parse-bench`](https://github.com/ophi-dev/antlr-rust-runtime/tree/main/tools/parse-bench) — Kotlin, C#, Java, Trino SQL — warm parses come out ahead of the Go runtime on every one, with Kotlin the outlier. It's a runnable harness (`run.py`, min/avg + ratios), so it's meant to be checked against your own grammars rather than taken as a universal claim. If it loses on a grammars-v4 grammar I'd genuinely want that data point.

On the two technical notes — both are real and both have answers today:

- **".interp doesn't represent actions."** Correct, and that's why there's a second path: `antlr4-rust-gen --actions embedded --grammar Foo.g4` splices the grammar's action/predicate bodies verbatim at their ATN coordinates (after `$`-attribute translation). That mode is exactly what drives the conformance harness through the action-bearing descriptors, so action grammars aren't limited to the metadata-only path. For portable idioms there's also a recognized-pattern library and `--sem-patterns`, with `--sem-unknown=error` to fail loud rather than silently no-op.
- **"`--grammar` doesn't seem to work."** That shouldn't happen — the flag is live. If you still have the invocation (grammar + command line), a one-line issue on the repo would be very welcome; I'll turn it around quickly. I'd rather fix the actual repro than guess.

And to your bigger suggestion — a complete `.interp`-with-actions that any target could consume — that's a genuinely interesting direction; happy to compare notes. For the record it's BSD-3, passes the upstream runtime-testsuite (357/357 discovered), builds on stable Rust, one runtime dependency. If it'd help your grammars-v4 testing, I can put together a build+run script in the [abnf_rust](https://github.com/kaby76/abnf_rust) spirit so it's a `cargo`-and-Bash exercise with no tool build.
